Support passing parse mode to parse/fingerprint functions

lcheruka opened the pull request:
Fixes #215

lcheruka committed:
Allow passing parse mode to parse/fingerprint functions
Added parseMode to json output for PLpgSQL_expr struct

seanlinsley commented on src/pg_query_parse.c:
```diff
@@ -42,7 +42,7 @@ PgQueryInternalParsetreeAndError pg_query_raw_parse(const char* input)
 	PG_TRY();
 	{
-		result.tree = raw_parser(input, RAW_PARSE_DEFAULT);
+		result.tree = raw_parser(input, mode);
```
I'm not sure what the fix should be, but there's a warning here:
```
compiling src/pg_query_parse.c
src/pg_query_parse.c:45:35: warning: implicit conversion from enumeration type 'PgQueryParseMode' to different enumeration type 'RawParseMode' [-Wenum-conversion]
                result.tree = raw_parser(input, mode);
                              ~~~~~~~~~~        ^~~~
```

seanlinsley commented on test/fingerprint_opts.c:
```diff
@@ -0,0 +1,54 @@
+		{
+			ret_code = -1;
+			printf("INVALID result for \"%s\" with %d mode\nexpected: \"%s\"\nactual: \"%s\"\nactual tokens: ", tests[i], mode, tests[i + 2], result.fingerprint_str);
+			pg_query_fingerprint_with_opts(tests[i], mode, true);
```
It looks like the `actual tokens: ` should be removed, as well as this `pg_query_fingerprint_with_opts` call. Or did you intend to do something more here?

lcheruka replied:
This was copied from the existing fingerprint.c. I believe the second call prints the actual tokens.

lcheruka committed:
- Fixed implicit conversion warning

lfittl commented:
Mainly commenting so I don't forget:

For the newly introduced `_opt` functions, I wonder if we should also consider the use cases requested in https://github.com/pganalyze/libpg_query/issues/50 here, i.e. to set GUCs that affect parsing.

Need to think through this more how it could be part of the API, but might be worth doing now so we don't have to break the `_opt` function API in the future.

lfittl committed:
Allow passing options to change GUCs that affect parser behaviour

Most importantly, this allows turning off standard_conforming_strings
to support legacy applications. It also adds support for turning
off backslash_quote and escape_string_warning.

These settings can be set by passing the following flags to the _opts
function variants. Use bitwise OR to combine flags, as well as set the
parse mode.

- PG_QUERY_DISABLE_BACKSLASH_QUOTE (backslash_quote = off)
- PG_QUERY_DISABLE_STANDARD_CONFORMING_STRINGS (standard_conforming_strings = off)
- PG_QUERY_DISABLE_ESCAPE_STRING_WARNING (escape_string_warning = off)

Fixes #50

lfittl commented:
> Need to think through this more how it could be part of the API, but might be worth doing now so we don't have to break the `_opt` function API in the future.

I've gone ahead and done that change now on this branch (thanks for allowing maintainer edits!), by utilizing an integer as a bitmask instead of a simple enum. From a usability perspective this is essentially the same as if you just care about the parse mode, but if you want to set the options you can just use bitwise OR to enable them.

Overall I think I'm happy with this now. @seanlinsley might be worth if you take another look to see if there are any concerns with that last commit, but otherwise I think we can merge.

lfittl committed:
Parse option tests: Add test that combines special mode with a GUC option

lfittl commented:
@lcheruka Thanks for the contribution! Merged with slight revisions to allow passing additional options beyond the parse mode.
